### ISSUE TITLE
Allow alternatives to XMLGregorianCalendar for xsd:dateTime

### ISSUE THREAD

**Issue by RomanIakovlev** [eed3si9n/scalaxb#376]
Currently there's no way to configure scalaxb to use anything but `XMLGregorianCalendar` for `xsd:dateTime` and other time-related types in generated classes. It would be convenient to use `java.time.*` classes instead, either by default or based on a config value.

**Comment by tOverney** [comment]
Thinking about having a go at it!
@eed3si9n what do you think of adding a settingKey `scalaxbUseJavaTime` defaulting to `false` ?

I would then change [places like this](https://github.com/eed3si9n/scalaxb/blob/f6b7819519630938119868a7d0c6cb376e3bbae1/cli/src/main/scala/scalaxb/compiler/xsd/XsTypeSymbol.scala#L164) to use that boolean and map the xs:date and other those to some new `BuiltInSimpleTypeSymbol` if the setting key is true otherwise keep the current mapping.

e.g.:

``` scala
object XsJavaTimeDate extends BuiltInSimpleTypeSymbol("java.time.LocalDate") {}

  ...
  case "date" if useJavaTime => XsJavaTimeDate
  case "date" => XsDate
  ...
```

**Comment by eed3si9n** [comment]
I'd say `scalaxbUseJavaTime` should default to hell yea.

**Comment by khajavi** [comment]
Hey @eed3si9n 
I sent this pull request https://github.com/eed3si9n/scalaxb/pull/542
This is my first try. Could you please review that pull request?
I have a question on what is the corresponding data type for xs:time in java.time.* package?

**Comment by tOverney** [comment]
Crap!
I guess I was not clear enough in my above post but I started to work on that too: https://github.com/eed3si9n/scalaxb/compare/master...tOverney:javaTime
Plann-ed/-ing to finish it tomorrow!

**Comment by tOverney** [comment]
The thing missing from my WIP is how to make [fromAny](https://github.com/eed3si9n/scalaxb/blob/master/cli/src/main/resources/scalaxb.scala.template#L476) work with the two possible types!

**Comment by khajavi** [comment]
Hey Tristan,
Great work but here are two notes about your works:
1. [LocalTime](https://github.com/eed3si9n/scalaxb/compare/master...tOverney:javaTime#diff-3cd286a586c649b63cdb9eb0689eeb8bde2d5ac354b0e493015832a2608926d5R217) does not correspond to the xs:time type because [xs:time has timeZoneOffset information](https://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/datatypes.html#time) in the xmlschema spec. So if we use LocalTime it loss the timeZoneOffset.
2. Also [LocalDate](https://github.com/eed3si9n/scalaxb/compare/master...tOverney:javaTime#diff-3cd286a586c649b63cdb9eb0689eeb8bde2d5ac354b0e493015832a2608926d5R206) is not correspond to the xs:DateTime. [xs:DateTime have timeZoneOffset](https://www.w3.org/TR/2012/REC-xmlschema11-2-20120405/datatypes.html#dateTime) but LocalDate doesn't.

This is why [I asked that question](https://github.com/eed3si9n/scalaxb/issues/376#issuecomment-723381957).

**Comment by tOverney** [comment]
Hey,
This is a valid point,
using LocalDate and LocalDateTime with the `ISO_*` DateTimeFormatter is fully compliant with the w3spec in terms of parsing and writing of values.

But indeed if you read a date that is not in your timezone you would write it back in your timezone (while still writting the same correct epoch instant).
And as far as I'm aware/concerned all apps I've worked with tend to work with in a single timezone with LocalDate everywhere;

I would argue that adding support for java.time.* is about convenience (as written in the initial post from @RomanIakovlev) and if one really needs to keep timezone data then that's where setting `useJavaTime = false` comes into place and you would still get classes that do not convert parsed values into your own timezone.

For the LocalTime I do see an issue since it's basically a 24h circular value and could completely break everything... I guess [OffsetTime](https://docs.oracle.com/javase/8/docs/api/java/time/OffsetTime.html) could do the trick, right?

**Comment by deenar** [comment]
Hi Tristan @tOverney. 

My 2 cents
>>using LocalDate and LocalDateTime with the ISO_* DateTimeFormatter is fully compliant with the w3spec in terms of parsing and writing of values
I was under the impression that al xs:dateTime, xs:date and xs:Time all have optional timezone elements, so just using  LocalDate and LocalDateTime would break the roundtrip that scalaxb and @eed3si9n triea hard to maintain. 

Using java.time.OffsetDateTime should support this and there is a toLocalDateTime and toLocalDate method on OffsetDateTime, so it is not too inconvenient for those working with dates from single timezones. 

Deenar

**Comment by eed3si9n** [comment]
Yes +1 on `OffsetDateTime` - https://docs.oracle.com/javase/8/docs/api/java/time/OffsetDateTime.html

This reminds me of a Twitter convo I had on this - https://twitter.com/jodastephen/status/882926344547520514

**Comment by tOverney** [comment]
Arguments make sense for `OffsetDateTime` for `xs:dateTime`!
but is using a type including a time element really a good idea for `xs:date`? I find this quite misleading

**Comment by eed3si9n** [comment]
https://www.w3.org/TR/xmlschema-2/#date

> The ·lexical space· of date consists of finite-length sequences of characters of the form: `'-'? yyyy '-' mm '-' dd zzzzzz?` where the date and optional timezone are represented exactly the same way as they are for dateTime. The first moment of the interval is that represented by: `'-' yyyy '-' mm '-' dd 'T00:00:00' zzzzzz?` and the least upper bound of the interval is the timeline point represented (noncanonically) by: `'-' yyyy '-' mm '-' dd 'T24:00:00' zzzzzz?`.

If we need to represent `2002-10-10+13`, I'm not sure if there are any other option than using `00:00:00` of `OffsetDateTime`.

**Comment by deenar** [comment]
There isn't an OffsetDate only type. I suspect this is really to do with the timezone offset, the LocalDate might be different given timezones. It is not different from the current state where XMLGregorianCalendar is used for all types.

**Comment by atais** [comment]
How about super flexible parameter:

`dateTypeClass` where user could provide a full path to the desired class. 
default: `javax.xml.datatype.XMLGregorianCalendar`
some would use `org.joda.time.DateTime` 
and others (like me) `java.util.String`.

You can read in the #543 why this would be important.

Do you think it could work?
